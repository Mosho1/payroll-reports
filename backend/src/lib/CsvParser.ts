import { Service } from "typedi";

@Service()
export class CsvParser {
    fromString<T extends { [index: string]: string }>(csvString: string): { id: string, entries: T[] } {
        const rows = csvString.trim().split('\n').map(row => row.split(','));
        const header = rows.splice(0, 1)[0];
        const footer = rows.splice(rows.length - 1, 1)[0];

        if (!footer) throw new Error('csv log must have a footer!')

        const entries = rows.map(row => header.reduce((entry, col, i) => {
            entry[col] = row[i];
            return entry;
        }, {} as T));

        const id = footer[1];

        return {id, entries};
    }
}