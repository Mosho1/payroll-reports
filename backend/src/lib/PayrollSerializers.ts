import { PayrollLog } from "../entity/PayrollLog";
import { PayrollEntry } from "../entity/PayrollEntry";
import { PayrollReport } from "../entity/PayrollReport";
import { Service, Inject } from "typedi";
import './CsvParser';
import { CsvParser } from "./CsvParser";
import { PayrollLogEntry } from "../types/PayrollLogEntry";
import { PayrollGroup } from "../entity/PayrollGroup";
import { InjectRepository } from "typeorm-typedi-extensions";
import { Repository } from "typeorm";
import { keyBy } from "lodash";

@Service()
export class PayrollSerializers {
    static cacheKeyPrefix = 'PayrollReport:';
    @Inject() private readonly csvParser: CsvParser;
    @InjectRepository(PayrollGroup) private readonly payrollGroupRepo: Repository<PayrollGroup>;

    payrollReportFromKeyValue(key: string, value: string) {
        const split = key.replace(PayrollSerializers.cacheKeyPrefix, '').split('-');
        return {
            amountPaid: Number(value),
            employee: Number(split[0]),
            year: Number(split[1]),
            month: Number(split[2]),
            isFirstHalf: split[3] === 'A',
        };
    }

    async getPayrollGroups() {
        // we can add caching here if needed
        const groupsArray = await this.payrollGroupRepo.find();
        return keyBy(groupsArray, 'id');
    }

    async payrollLogToKeyValues(log: PayrollLog) {
        const payrollGroups = await this.getPayrollGroups();
        let map: { [index: string]: number } = {};
        for (const entry of log.entries!) {
            const year = entry.date.getFullYear();
            const month = entry.date.getMonth();
            const isFirstHalf = isFirstHalfOfMonth(entry.date);
            const key = `${PayrollSerializers.cacheKeyPrefix}${entry.employee}-${year}-${month}-${isFirstHalf ? 'A' : 'B'}`;

            const payrollGroup = payrollGroups[entry.payrollGroup!.id];

            if (!payrollGroup) throw new Error(`Can't find payroll group with id ${entry.payrollGroup!.id}`);

            map[key] = map[key] || 0;
            map[key] += entry.hoursWorked * payrollGroup.hourlyRate;
        }
        return map;
    }

    getDate(strDate: string) {
        const [day, month, year] = strDate.split('/');
        return new Date(
            Number(year),
            Number(month) - 1,
            Number(day)
        );
    }

    csvLogToPayrollLog(csvString: string) {
        const { id, entries } = this.csvParser.fromString<PayrollLogEntry>(csvString);
        const numberId = Number(id);
        
        if (Number.isNaN(numberId)) throw new Error('Report id must be a number');

        let log = new PayrollLog();
        log.id = numberId;
        log.entries = entries.map((entry): PayrollEntry => {
            const payrollEntry = {
                date: this.getDate(entry['date']),
                employee: Number(entry['employee id']),
                hoursWorked: Number(entry['hours worked']),
                payrollLog: log,
                payrollGroup: new PayrollGroup(entry['job group'])
            };
            return payrollEntry;
        })

        return log;
    }

}

const isFirstHalfOfMonth = (date: Date) => date.getDate() <= 15;
