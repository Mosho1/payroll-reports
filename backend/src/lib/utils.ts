import { promisify } from 'util';

export const promisifyMethod = <U, T, K extends keyof T>(object: T, methodName: K) => {
    const method = object[methodName];
    if (typeof method !== 'function') throw new Error(`Key ${method} is not a method`)
    return promisify<U>(method.bind(object));
};