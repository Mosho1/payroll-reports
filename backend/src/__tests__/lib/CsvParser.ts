import { CsvParser } from "../../lib/CsvParser";
import {promises as fs} from 'fs';
import { join } from "path";
import Container from "typedi";

describe('PayrollData', () => {
    it('parses a csv string', async () => {
        const csvParser = Container.get(CsvParser);
        const csv = await fs.readFile(join(__dirname, 'sample.csv'));
        const report = csvParser.fromString(csv.toString());
        expect(report).toMatchSnapshot();
    });
});