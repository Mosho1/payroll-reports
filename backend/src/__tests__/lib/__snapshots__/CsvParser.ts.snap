// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayrollData parses a csv string 1`] = `
Object {
  "entries": Array [
    Object {
      "date": "14/11/2016",
      "employee id": "1",
      "hours worked": "7.5",
      "job group": "A",
    },
    Object {
      "date": "9/11/2016",
      "employee id": "2",
      "hours worked": "4",
      "job group": "B",
    },
    Object {
      "date": "10/11/2016",
      "employee id": "2",
      "hours worked": "4",
      "job group": "B",
    },
    Object {
      "date": "9/11/2016",
      "employee id": "3",
      "hours worked": "11.5",
      "job group": "A",
    },
    Object {
      "date": "8/11/2016",
      "employee id": "3",
      "hours worked": "6",
      "job group": "A",
    },
    Object {
      "date": "11/11/2016",
      "employee id": "3",
      "hours worked": "3",
      "job group": "A",
    },
    Object {
      "date": "2/11/2016",
      "employee id": "3",
      "hours worked": "6",
      "job group": "A",
    },
    Object {
      "date": "3/11/2016",
      "employee id": "2",
      "hours worked": "12",
      "job group": "B",
    },
    Object {
      "date": "4/11/2016",
      "employee id": "2",
      "hours worked": "11",
      "job group": "B",
    },
    Object {
      "date": "6/11/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "21/11/2016",
      "employee id": "1",
      "hours worked": "6",
      "job group": "A",
    },
    Object {
      "date": "22/11/2016",
      "employee id": "1",
      "hours worked": "5",
      "job group": "A",
    },
    Object {
      "date": "23/11/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "24/11/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "25/11/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "14/12/2016",
      "employee id": "1",
      "hours worked": "7.5",
      "job group": "A",
    },
    Object {
      "date": "9/12/2016",
      "employee id": "2",
      "hours worked": "4",
      "job group": "B",
    },
    Object {
      "date": "10/12/2016",
      "employee id": "2",
      "hours worked": "4",
      "job group": "B",
    },
    Object {
      "date": "9/12/2016",
      "employee id": "3",
      "hours worked": "11.5",
      "job group": "A",
    },
    Object {
      "date": "8/12/2016",
      "employee id": "3",
      "hours worked": "6",
      "job group": "A",
    },
    Object {
      "date": "12/11/2016",
      "employee id": "3",
      "hours worked": "3",
      "job group": "A",
    },
    Object {
      "date": "2/12/2016",
      "employee id": "3",
      "hours worked": "6",
      "job group": "A",
    },
    Object {
      "date": "3/12/2016",
      "employee id": "2",
      "hours worked": "12",
      "job group": "B",
    },
    Object {
      "date": "4/12/2016",
      "employee id": "2",
      "hours worked": "11",
      "job group": "B",
    },
    Object {
      "date": "6/12/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "21/12/2016",
      "employee id": "1",
      "hours worked": "6",
      "job group": "A",
    },
    Object {
      "date": "22/12/2016",
      "employee id": "1",
      "hours worked": "5",
      "job group": "A",
    },
    Object {
      "date": "23/12/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "24/12/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "25/12/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
    Object {
      "date": "23/2/2015",
      "employee id": "4",
      "hours worked": "5",
      "job group": "A",
    },
    Object {
      "date": "24/2/2016",
      "employee id": "4",
      "hours worked": "5",
      "job group": "B",
    },
  ],
  "id": "43",
}
`;
